### PR TITLE
C++: Don't strip specifiers away in `TFinalParameterUse`

### PR DESCRIPTION
**Pull request by MathiasVP**
This is a preparatory change to an upcoming PR I'm doing related to the "automatically block flow out of modelled functions" work. It won't really have any visible impact yet, so it should be a totally safe change.

The core of the change is that we were incorrectly calling `getUnspecifiedType` instead of `getUnderlyingType` when checking whether a parameter could be written to. Calling `getUnspecifiedType` is wrong since, in addition to resolve typedefs, it also strips off `const` specifiers.

Instead, we should be calling `getUnderlyingType` since that _only_ resolves typedefs.

DCA looks uneventful (as expected).

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll [hunk 1]
```diff
@@ -452,7 +452,7 @@ private module IsModifiableAtImpl {
   private predicate impl(CppType cppType, int indirectionIndex) {
     exists(Type pointerType, Type base |
       isUnderlyingIndirectionType(pointerType) and
-      cppType.hasUnderlyingType(pointerType, _) and
+      cppType.hasUnderlyingType(pointerType, false) and
```
Comment (MathiasVP) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll:455]: This was a small drive-by optimiation I did: Since we only ever care about non-glvalue types in this predicate there's no need to get both the glvalue _and_ the prvalue version of the type as we compute this pipeline

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll [hunk 2]
```diff
@@ -172,11 +172,13 @@ private predicate isGlobalDefImpl(
...
-private predicate unspecifiedTypeIsModifiableAt(Type unspecified, int indirectionIndex) {
-  indirectionIndex = [1 .. getIndirectionForUnspecifiedType(unspecified).getNumberOfIndirections()] and
+private predicate underlyingTypeIsModifiableAt(Type underlying, int indirectionIndex) {
+  indirectionIndex =
+    [1 .. getIndirectionForUnspecifiedType(underlying.getUnspecifiedType())
+          .getNumberOfIndirections()] and
   exists(CppType cppType |
-    cppType.hasUnspecifiedType(unspecified, _) and
-    isModifiableAt(cppType, indirectionIndex + 1)
+    cppType.hasUnderlyingType(underlying, false) and
```
Comment (MathiasVP) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll:180]: The `false` here is a similar drive-by optimization just like [here](https://github.com/github/codeql/pull/15626/files#r1490990071)

**Comment by MathiasVP** [comment]
DCA looks uneventful (as expected)

**Review by jketema — APPROVED** [review]
